cpp/dolfinx/fem/utils.h

```diff
@@ -663,21 +655,39 @@ void pack_coefficients(const Form<T>& form, IntegralType integral_type, int id,
 
   if (!coefficients.empty())
   {
-    xtl::span<const std::uint32_t> cell_info
-        = impl::get_cell_orientation_info(coefficients);
-
     switch (integral_type)
     {
     case IntegralType::cell:
     {
-      auto fetch_cell = [](auto entity) { return entity; };
       const std::vector<std::int32_t>& cells = form.cell_domains(id);
       // Iterate over coefficients
       for (std::size_t coeff = 0; coeff < coefficients.size(); ++coeff)
       {
-        impl::pack_coefficient_entity(c, cstride, *coefficients[coeff],
-                                      cell_info, cells, fetch_cell,
-                                      offsets[coeff]);
+        // Get cell info for coefficient (with respect to coefficient mesh)
+        xtl::span<const std::uint32_t> cell_info
+            = impl::get_cell_orientation_info(*coefficients[coeff]);
+
+        // If the coefficient is defined over a different mesh to the
+        // integration domain mesh, we must map entities from the
+        // integration domain mesh to the coefficient mesh to pack them
+        // properly
+        auto coeff_mesh = coefficients[coeff]->function_space()->mesh();
+        if (coeff_mesh != form.mesh())
+        {
+          auto fetch_cell
+              = [&entity_map = form.entity_maps().at(coeff_mesh)](auto entity)
+          { return entity_map[entity]; };
```

Comment:
Is ` entity_map` a `std::map`? If it is we should use `find` rather than `[]` for the look up.